crates/uv-resolver/src/pubgrub/priority.rs

```diff
@@ -1,34 +1,105 @@
 use std::cmp::Reverse;
 
+use pubgrub::range::Range;
 use rustc_hash::FxHashMap;
 
+use pep440_rs::Version;
 use uv_normalize::PackageName;
 
 use crate::pubgrub::package::PubGrubPackage;
 
 #[derive(Debug, Default)]
-pub(crate) struct PubGrubPriorities(FxHashMap<PackageName, usize>);
+pub(crate) struct PubGrubPriorities(FxHashMap<PackageName, PubGrubPriority>);
 
 impl PubGrubPriorities {
-    /// Add a package to the priority map.
-    pub(crate) fn add(&mut self, package: PackageName) {
-        let priority = self.0.len();
-        self.0.entry(package).or_insert(priority);
+    /// Add a [`PubGrubPackage`] to the priority map.
+    pub(crate) fn insert(&mut self, package: &PubGrubPackage, version: &Range<Version>) {
+        let next = self.0.len();
+        match package {
+            PubGrubPackage::Root(_) => {}
+            PubGrubPackage::Python(_) => {}
+            PubGrubPackage::Package(name, _, None) => {
+                match self.0.entry(name.clone()) {
+                    std::collections::hash_map::Entry::Occupied(mut entry) => {
+                        let index = match entry.get() {
+                            PubGrubPriority::Singleton(Reverse(index)) => *index,
+                            PubGrubPriority::Unconstrained(Reverse(index)) => *index,
+                            PubGrubPriority::Constrained(Reverse(index)) => *index,
+                            PubGrubPriority::DirectUrl(Reverse(index)) => *index,
+                            PubGrubPriority::Root => unreachable!(),
+                        };
+
+                        // Update the priority, but preserve the original index.
```

Comment:
Yeah that's true. I could change it to always take the max priority between new and existing? It will sometimes be wrong if we backtrack, but it's just a heuristic so it's ok.

Reply:
max of old and new sounds like a good heuristic